core/src/main/java/io/parsingdata/metal/expression/value/reference/Ref.java

```diff
@@ -49,37 +50,32 @@
 
     public final T reference;
     public final Predicate<ParseValue> predicate;
-    public final ValueExpression limit;
+    public final SingleValueExpression limit;
 
-    private Ref(final T reference, final Predicate<ParseValue> predicate, final ValueExpression limit) {
+    private Ref(final T reference, final Predicate<ParseValue> predicate, final SingleValueExpression limit) {
         this.reference = checkNotNull(reference, "reference");
         this.predicate = checkNotNull(predicate, "predicate");
         this.limit = limit;
     }
 
     public static class NameRef extends Ref<String> {
         public NameRef(final String reference) { this(reference, null); }
-        public NameRef(final String reference, final ValueExpression limit) { super(reference, value -> value.matches(reference), limit); }
+        public NameRef(final String reference, final SingleValueExpression limit) { super(reference, value -> value.matches(reference), limit); }
     }
 
     public static class DefinitionRef extends Ref<Token> {
         public DefinitionRef(final Token reference) { this(reference, null); }
-        public DefinitionRef(final Token reference, final ValueExpression limit) { super(reference, value -> value.definition.equals(reference), limit); }
+        public DefinitionRef(final Token reference, final SingleValueExpression limit) { super(reference, value -> value.definition.equals(reference), limit); }
     }
 
     @Override
     public ImmutableList<Value> eval(final ParseState parseState, final Encoding encoding) {
         if (limit == null) {
             return evalImpl(parseState, NO_LIMIT);
         }
-        final ImmutableList<Value> evaluatedLimit = limit.eval(parseState, encoding);
-        if (evaluatedLimit.size != 1) {
-            throw new IllegalArgumentException("Limit must evaluate to a single non-empty value.");
-        }
-        if (evaluatedLimit.head.equals(NOT_A_VALUE)) {
-            return ImmutableList.create(NOT_A_VALUE);
-        }
-        return evalImpl(parseState, evaluatedLimit.head.asNumeric().intValueExact());
+        return limit.evalSingle(parseState, encoding)
+            .map(limitValue -> limitValue.equals(NOT_A_VALUE) ? ImmutableList.create(NOT_A_VALUE) : evalImpl(parseState, limitValue.asNumeric().intValueExact()))
```

Comment:
The error message in combination with this logic confuses me. In this case, `limitValue` could be `NOT_A_VALUE`, but the error message states that the limit must evaluate to a non-empty value. But we guard that case here?
So, either the error message is incorrect and we can rewrite this as:
```
limit.evalSingle(parseState, encoding)
    .filter(not(limitValue -> limitValue.equals(NOT_A_VALUE)))
    .map(limitValue -> evalImpl(parseState, limitValue.asNumeric().intValueExact()))
    .orElseThrow(() -> new IllegalArgumentException("Ref must evaluate to a non-empty value."))
```
or we can use `orElseGet` instead:
```
limit.evalSingle(parseState, encoding)
    .filter(not(limitValue -> limitValue.equals(NOT_A_VALUE)))
    .map(limitValue -> evalImpl(parseState, limitValue.asNumeric().intValueExact()))
    .orElseGet(() -> ImmutableList.create(NOT_A_VALUE))
```
You may even consider adding an extra `map` step to first convert `limitValue` to int value, before passing it to `evalImpl`, because that may improve readability:
```
.map(limitValue -> limitValue.asNumeric().intValueExact())
```
This latter is just a thought. :)

Reply:
The problem is that there are three cases that must be handled based on what `limit` evaluates to:
1. An empty `Optional`, in which case an exception is thrown.
2. `NOT_A_VALUE`, in which case we return a list with that in it.
3. Something else, in which case we pass that to `evalImpl(...)` and return the result.
What you propose always handles only two of these cases I think, or do you propose to do something else next to this?

Reply:
After discussion, we postpone cleaning this until later.